Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm

```diff
@@ -1933,6 +1938,107 @@ void testSuspendServiceWorkerProcessBasedOnClientProcesses(UseSeparateServiceWor
+{
+    RetainPtr<NSBundle> main = [NSBundle mainBundle];
+    RetainPtr<NSString> fileName = [NSString stringWithCString:filename.c_str() encoding:NSUTF8StringEncoding];
+    RetainPtr<NSString> extension = [NSString stringWithCString:type.c_str() encoding:NSUTF8StringEncoding];
+    RetainPtr<NSString> resourcePath = [main pathForResource:fileName.get() ofType:extension.get() inDirectory:@"TestWebKitAPI.resources"];
+
+    RetainPtr<NSData> fileContent = [NSData dataWithContentsOfFile:resourcePath.get()];
+    if (!fileContent)
+        return "";
+    RetainPtr<NSString> fileContentString = adoptNS([[NSString alloc] initWithBytes:fileContent.get().bytes length:fileContent.get().length encoding:NSUTF8StringEncoding]);
+    return std::string([fileContentString UTF8String]);
+}
+
+TEST(ServiceWorkers, WrapUnwrapCryptoKeyFromServiceWorker)
+{
+    [WKWebsiteDataStore _allowWebsiteDataRecordsForAllOrigins];
+
+    // Start with a clean slate data store
+    [[WKWebsiteDataStore defaultDataStore] removeDataOfTypes:[WKWebsiteDataStore allWebsiteDataTypes] modifiedSince:[NSDate distantPast] completionHandler:^() {
+        done = true;
+    }];
+    TestWebKitAPI::Util::run(&done);
+    done = false;
+
+    RetainPtr<WKWebViewConfiguration> configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
+
+
+    RetainPtr<SWCryptoKeyMessageHandler> messageHandler = adoptNS([[SWCryptoKeyMessageHandler alloc] init]);
+    [[configuration userContentController] addScriptMessageHandler:messageHandler.get() name:@"sw"];
+
+    auto mainBytesLocal = readFile("WrapUnwrapCryptoKeyServiceWorker", "html");
+    auto serviceWorkerBytesLocal = readFile("indexeddb-cryptokey-put-get-worker", "js");
+    ASSERT_GT(mainBytesLocal.length(), 0ul);
+    ASSERT_GT(serviceWorkerBytesLocal.length(), 0ul);
+    TestWebKitAPI::HTTPServer server({
+        { "/"_s, { ASCIILiteral::fromLiteralUnsafe(mainBytesLocal.c_str()) } },
+        { "/sw.js"_s, { { { "Content-Type"_s, "application/javascript"_s } }, ASCIILiteral::fromLiteralUnsafe(serviceWorkerBytesLocal.c_str()) } },
+    });
+
+
+    RetainPtr<WKWebView> webView = adoptNS([[WKWebView alloc] initWithFrame:NSMakeRect(0, 0, 800, 600) configuration:configuration.get()]);
+
+    auto delegate = adoptNS([[TestSWAsyncNavigationDelegate alloc] init]);
+    [webView setNavigationDelegate:delegate.get()];
+    [webView setUIDelegate:delegate.get()];
+    done = false;
+
+    // Run Test without Delegate.
+    [webView loadRequest:server.request()];
+    TestWebKitAPI::Util::run(&done);
+    done = false;
+
+    // Install delegate for masterKey.
+    auto dataStoreDelegate = adoptNS([[CryptoKeyWebsiteDataStoreDelegate alloc] init]);
+    [[configuration websiteDataStore] set_delegate:dataStoreDelegate.get()];
+
+    [webView loadRequest:server.request()];
+    TestWebKitAPI::Util::run(&done);
+    done = false;
+
+    // assert check masterKey fetch
+    ASSERT_EQ(15ul, [dataStoreDelegate fetchCount]);
```

Comment:
It seems this test mainly ensures the delegate method is called and the rest is the same as the layout test, so I don't think we need this API test and the layout test is enough.
If you want to test the delegate method, we should probably implement it WebKitTestRunner to make the test not use the default master key.

Reply:
I need an API test to make sure of two things in order to ensure that this is not causing a regression:
1. Make sure that this change does not crash the process until we add the delegate implementation in Safari.
2. Install the delegate and be able to assert that a C++ delegate API is called.

Reply:
> If you want to test the delegate method, we should probably implement it WebKitTestRunner to make the test not use the default master key.

How do you propose I do that ? Is there an example of such a tweak that can control C++ layer behavior from JavaScript ?